models/issue_label.go

```diff
@@ -632,6 +632,8 @@ func HasIssueLabel(issueID, labelID int64) bool {
 	return hasIssueLabel(x, issueID, labelID)
 }
 
+// newIssueLabel this function creates a new label it does not check if the label is valid for the issue
+// YOU MUST CHECK THIS BEFORE THIS FUNCTION
```

Comment:
```diff
diff --git a/models/issue_label.go b/models/issue_label.go
index 54b286fe7..2da0d2f20 100644
--- a/models/issue_label.go
+++ b/models/issue_label.go
@@ -632,7 +632,43 @@ func HasIssueLabel(issueID, labelID int64) bool {
        return hasIssueLabel(x, issueID, labelID)
 }

+// ErrAddIssueLabelNotAllowed represents the error adding the label to issue not allowed
+type ErrAddIssueLabelNotAllowed struct {
+       LabelID     int64
+       LabelRepoID int64
+       LabelOrgID  int64
+       IssueID     int64
+       IssueRepoID int64
+       IssueOrgID  int64
+}
+
+func (err ErrAddIssueLabelNotAllowed) Error() string {
+       return fmt.Sprintf("Add label %d[%d,%d] to issue/pr %d failed because it's not belong respository %d or orgnization %d",
+               err.LabelID, err.LabelRepoID, err.LabelOrgID,
+               err.IssueID,
+               err.IssueRepoID,
+               err.IssueOrgID,
+       )
+}
+
 func newIssueLabel(e *xorm.Session, issue *Issue, label *Label, doer *User) (err error) {
+       if label.RepoID > 0 && label.RepoID != issue.RepoID {
+               return ErrAddIssueLabelNotAllowed{
+                       LabelID:     label.ID,
+                       LabelRepoID: label.RepoID,
+                       LabelOrgID:  label.OrgID,
+                       IssueID:     issue.ID,
+                       IssueRepoID: issue.RepoID,
+               }
+       } else if label.OrgID > 0 && label.OrgID != issue.Repo.OwnerID {
+               return ErrAddIssueLabelNotAllowed{
+                       LabelID:     label.ID,
+                       LabelRepoID: label.RepoID,
+                       LabelOrgID:  label.OrgID,
+                       IssueID:     issue.ID,
+                       IssueOrgID:  issue.Repo.OwnerID,
+               }
+       }
        if _, err = e.Insert(&IssueLabel{
                IssueID: issue.ID,
                LabelID: label.ID,
```

Reply:
That's gonna need to be added to a lot more places as there are lots of other places we silently drop these labels.